File: lib/personas.ts
import { query, queryOne } from './db'
import type { Persona, PersonaInput } from './types'

const COLS = `id, slug, name, title, description, avatar_url,
  system_prompt, analysis_prompt, analysis_schema,
  profile, profile_draft, profile_meta, profile_updated_at, consent_ack,
  created_at, updated_at`

/** Build a url-friendly slug from a name. */
export function slugify(name: string): string {
  return name
    .toLowerCase()
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '') // strip accents
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 60)
}

export async function listPersonas(): Promise<Persona[]> {
  return query<Persona>(`select ${COLS} from personas order by created_at asc`)
}

export async function getPersona(id: string): Promise<Persona | null> {
  return queryOne<Persona>(`select ${COLS} from personas where id = $1`, [id])
}

/** Ensure a slug is unique by appending -2, -3, ... when needed. */
async function uniqueSlug(base: string): Promise<string> {
  const root = base || 'persona'
  let slug = root
  let n = 1
  // eslint-disable-next-line no-constant-condition
  while (true) {
    const existing = await queryOne(`select 1 from personas where slug = $1`, [slug])
    if (!existing) return slug
    n += 1
    slug = `${root}-${n}`
  }
}

export async function createPersona(input: PersonaInput): Promise<Persona> {
  const slug = await uniqueSlug(slugify(input.name))
  const row = await queryOne<Persona>(
    `insert into personas
       (slug, name, title, description, avatar_url, system_prompt, analysis_prompt, analysis_schema,
        profile, profile_meta, consent_ack)
     values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
     returning ${COLS}`,
    [
      slug,
      input.name,
      input.title ?? null,
      input.description ?? null,
      input.avatar_url ?? null,
      input.system_prompt,
      input.analysis_prompt ?? null,
      input.analysis_schema ? JSON.stringify(input.analysis_schema) : null,
      input.profile ? JSON.stringify(input.profile) : null,
      input.profile_meta ? JSON.stringify(input.profile_meta) : null,
      input.consent_ack ?? false,
    ]
  )
  return row as Persona
}

export async function updatePersona(
  id: string,
  patch: Partial<PersonaInput>
): Promise<Persona> {
  // Build a dynamic SET clause from the provided fields only.
  const fields: string[] = []
  const values: unknown[] = []
  const set = (col: string, val: unknown) => {
    values.push(val)
    fields.push(`${col} = $${values.length}`)
  }

  if (patch.name !== undefined) set('name', patch.name)
  if (patch.title !== undefined) set('title', patch.title)
  if (patch.description !== undefined) set('description', patch.description)
  if (patch.avatar_url !== undefined) set('avatar_url', patch.avatar_url)
  if (patch.system_prompt !== undefined) set('system_prompt', patch.system_prompt)
  if (patch.analysis_prompt !== undefined) set('analysis_prompt', patch.analysis_prompt)
  if (patch.analysis_schema !== undefined) {
    set('analysis_schema', patch.analysis_schema ? JSON.stringify(patch.analysis_schema) : null)
  }
  if (patch.consent_ack !== undefined) set('consent_ack', patch.consent_ack)

  // Profile fields are JSONB. Only a change to the APPROVED profile bumps
  // profile_updated_at (the optimistic-lock token); saving a draft does not.
  if (patch.profile !== undefined) {
    set('profile', patch.profile ? JSON.stringify(patch.profile) : null)
    fields.push(`profile_updated_at = now()`)
  }
  if (patch.profile_draft !== undefined) {
    set('profile_draft', patch.profile_draft ? JSON.stringify(patch.profile_draft) : null)
  }
  if (patch.profile_meta !== undefined) {
    set('profile_meta', patch.profile_meta ? JSON.stringify(patch.profile_meta) : null)
  }

  if (fields.length === 0) {
    const current = await getPersona(id)
    if (!current) throw new Error('Persona não encontrada.')
    return current
  }

  values.push(id)
  const row = await queryOne<Persona>(
    `update personas set ${fields.join(', ')} where id = $${values.length} returning ${COLS}`,
    values
  )
  if (!row) throw new Error('Persona não encontrada.')
  return row
}

export async function deletePersona(id: string): Promise<void> {
  await query(`delete from personas where id = $1`, [id])
}

/** Count knowledge chunks for a persona (shown in the UI). */
export async function countDocuments(personaId: string): Promise<number> {
  const row = await queryOne<{ count: string }>(
    `select count(*)::int as count from documents where persona_id = $1`,
    [personaId]
  )
  return Number(row?.count ?? 0)
}

File: backend/app/services/personas.py
"""Persona CRUD. Port of lib/personas.ts."""
from __future__ import annotations

import re
import unicodedata
from typing import Any

from ..db import query, query_one

COLS = """id, slug, name, title, description, avatar_url,
  system_prompt, analysis_prompt, analysis_schema,
  profile, profile_draft, profile_meta, profile_updated_at, consent_ack,
  created_at, updated_at"""


def slugify(name: str) -> str:
    """Build a url-friendly slug from a name."""
    nfkd = unicodedata.normalize("NFD", name.lower())
    no_accents = "".join(c for c in nfkd if unicodedata.category(c) != "Mn")
    slug = re.sub(r"[^a-z0-9]+", "-", no_accents)
    slug = re.sub(r"^-+|-+$", "", slug)
    return slug[:60]


async def list_personas() -> list[dict[str, Any]]:
    return await query(f"select {COLS} from personas order by created_at asc")


async def get_persona(persona_id: str) -> dict[str, Any] | None:
    return await query_one(f"select {COLS} from personas where id = $1::uuid", [persona_id])


async def _unique_slug(base: str) -> str:
    """Ensure a slug is unique by appending -2, -3, ... when needed."""
    root = base or "persona"
    slug = root
    n = 1
    while True:
        existing = await query_one("select 1 from personas where slug = $1", [slug])
        if not existing:
            return slug
        n += 1
        slug = f"{root}-{n}"


async def create_persona(data: dict[str, Any]) -> dict[str, Any]:
    slug = await _unique_slug(slugify(data["name"]))
    row = await query_one(
        f"""insert into personas
              (slug, name, title, description, avatar_url, system_prompt, analysis_prompt, analysis_schema,
               profile, profile_meta, consent_ack)
            values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
            returning {COLS}""",
        [
            slug,
            data["name"],
            data.get("title"),
            data.get("description"),
            data.get("avatar_url"),
            data["system_prompt"],
            data.get("analysis_prompt"),
            data.get("analysis_schema"),
            data.get("profile"),
            data.get("profile_meta"),
            data.get("consent_ack", False),
        ],
    )
    assert row is not None
    return row


# Columns that may be patched and whether changing them bumps profile_updated_at.
async def update_persona(persona_id: str, patch: dict[str, Any]) -> dict[str, Any]:
    """Build a dynamic SET clause from the provided fields only.

    `patch` uses the presence of a key (even with value None) to mean "set this".
    """
    fields: list[str] = []
    values: list[Any] = []

    def setcol(col: str, val: Any) -> None:
        values.append(val)
        fields.append(f"{col} = ${len(values)}")

    for col in (
        "name", "title", "description", "avatar_url",
        "system_prompt", "analysis_prompt", "analysis_schema", "consent_ack",
    ):
        if col in patch:
            setcol(col, patch[col])

    # Profile fields are JSONB. Only a change to the APPROVED profile bumps
    # profile_updated_at (the optimistic-lock token); saving a draft does not.
    if "profile" in patch:
        setcol("profile", patch["profile"])
        fields.append("profile_updated_at = now()")
    if "profile_draft" in patch:
        setcol("profile_draft", patch["profile_draft"])
    if "profile_meta" in patch:
        setcol("profile_meta", patch["profile_meta"])

    if not fields:
        current = await get_persona(persona_id)
        if not current:
            raise ValueError("Persona não encontrada.")
        return current

    values.append(persona_id)
    row = await query_one(
        f"update personas set {', '.join(fields)} where id = ${len(values)}::uuid returning {COLS}",
        values,
    )
    if not row:
        raise ValueError("Persona não encontrada.")
    return row


async def delete_persona(persona_id: str) -> None:
    await query("delete from personas where id = $1::uuid", [persona_id])


async def count_documents(persona_id: str) -> int:
    row = await query_one(
        "select count(*)::int as count from documents where persona_id = $1::uuid",
        [persona_id],
    )
    return int(row["count"]) if row else 0
